### PR TITLE
Fix for some invalid server behaviors when a client is aborting a request

### PR DESCRIPTION
**Pull request by lorban**
Fix three bugs that were discovered at the same time:

1) Fix `java.lang.IllegalArgumentException: demand pending` warning log. When a ReadListener is registered on a EE10 ServletInputStream after startAsync() has been called, and the client abruptly closes the connection, the server is left with a pending demand and trying to complete the `HttpChannelState`'s callback. The latter expects no pending demand otherwise `IllegalArgumentException: demand pending` is thrown. We should fail the callback with the IllegalArgumentException instead of throwing it.

2) Add the necessary wiring so that QUIC can fail its session when a HTTP 3 client aborts. A QUIC reset packet is immediately sent to the server which is supposed to react to it; when the QUIC endpoint has fill interest, this reset is passed to the HTTP 3 layer which does the right thing; but when the endpoint has no fill interest, the reset is incorrectly ignored.

3) Abort instead of throwing when the exception handler in EE10 `ServletChannel.handle()` throws an exception to avoid infinitely looping and adding an infinite amount of suppressed stack traces when the SEND_ERROR action itself throws.

See #11631

### CODE REVIEW

**Comment by lorban** [comment]
I could not find the cause of the `HttpFields` corruption that makes `HttpFields.Mutable.remove()` throw NPE, but in the process of trying to reproduce the issue I found two other bugs:

 - making the client abort while the servlet has started async, has registered an input listener and isn't processing makes the EE10 servlet state machine fail with ISE.
 - making the client abort a HTTP3 stream is ignored by the server.

I've committed fixes for those.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java [hunk 1]
```diff
@@ -15,23 +15,199 @@
...
+
+        AtomicReference<Result> resultRef = new AtomicReference<>();
+        try (AsyncRequestContent content = new AsyncRequestContent())
+        {
+            Request request = client.newRequest(newURI(transport))
+                .method("POST")
+                .timeout(5, TimeUnit.SECONDS)
+                .body(content);
+            request.send(resultRef::set);
+            servletDoneLatch.await();
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Please wait with a timeout and assert.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java [hunk 2]
```diff
@@ -15,23 +15,199 @@
...
+            request.abort(new ArithmeticException());
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: I don't think this line is relevant?

Let's try a bit harder to test that the server side `ServletChannel` state machine does not enter invalid states.
Can we notice the invalid states by registering listeners or similar?

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java [hunk 3]
```diff
@@ -15,23 +15,199 @@
...
+
+        AtomicReference<Result> resultRef = new AtomicReference<>();
+        try (AsyncRequestContent content = new AsyncRequestContent(ByteBuffer.allocate(16)))
+        {
+            Request request = client.newRequest(newURI(transport))
+                .method("POST")
+                .timeout(5, TimeUnit.SECONDS)
+                .body(content);
+            request.send(resultRef::set);
+            servletDoneLatch.await();
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Please wait with a timeout and assert.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java [hunk 4]
```diff
@@ -15,23 +15,199 @@
...
+            request.abort(new ArithmeticException());
+        }
+
+        await().pollDelay(1, TimeUnit.MILLISECONDS).atMost(5, TimeUnit.SECONDS).until(resultRef::get, not(nullValue()));
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Remove pollDelay().

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java [hunk 5]
```diff
@@ -15,23 +15,199 @@
...
+                .body(content);
+            request.send(resultRef::set);
+            servletDoneLatch.await();
+            request.abort(new ArithmeticException());
+        }
+
+        assertTrue(errorDoneLatch.await(5, TimeUnit.SECONDS));
+        assertThat(serverError.get(), instanceOf(EofException.class));
+
+        await().pollDelay(1, TimeUnit.MILLISECONDS).atMost(5, TimeUnit.SECONDS).until(resultRef::get, not(nullValue()));
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Remove pollDelay().

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java [hunk 6]
```diff
@@ -15,23 +15,199 @@
...
+        Result result = resultRef.get();
+        assertThat(result.getFailure(), instanceOf(ArithmeticException.class));
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Can you explicitly assert whether you expect a 567 response or not?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java (hunk 1 above)
Comment (lorban) [reply jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Done.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java (hunk 3 above)
Comment (lorban) [reply jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Done.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java (hunk 4 above)
Comment (lorban) [reply jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Done.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java (hunk 5 above)
Comment (lorban) [reply jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Done.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java (hunk 6 above)
Comment (lorban) [reply jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: I've added an assertion for the status code.

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java (hunk 2 above)
Comment (lorban) [reply jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/RequestReaderTest.java]: Yes, aborting the request is irrelevant as it's the fact that the server has a pending demand that triggers the bug. I've removed that call.

It's actually possible to get the `IllegalStateException` by inserting a handler on the server and wrapping the handler's callback to capture the failure. I've modified the test to do that and assert that no failure is reported.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java [hunk 7]
```diff
@@ -1455,9 +1455,9 @@ public void succeeded()
                 // We are being tough on handler implementations and expect them
                 // to not have pending operations when calling succeeded or failed.
                 if (httpChannelState._onContentAvailable != null)
-                    throw new IllegalStateException("demand pending");
+                    failure = ExceptionUtil.combine(failure, new IllegalStateException("demand pending"));
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:1457]: You need to change the comment above.  We are now being less tough as we are converting a call to `succeeded` with pending demand/write into a call to `failure`.

Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:1457]: Right, I've modified the comment accordingly.

**Review by sbordet — APPROVED** [review]

**Review by gregw — APPROVED** [review]

**Pull request merged**